include/fiction/algorithms/physical_design/hexagonalization.hpp

```diff
@@ -0,0 +1,200 @@
+//
+// Created by simon on 14.04.23.
+//
+
+#ifndef FICTION_HEXAGONALIZATION_HPP
+#define FICTION_HEXAGONALIZATION_HPP
+
+#include "fiction/traits.hpp"
+#include "fiction/types.hpp"
+
+namespace fiction
+{
+/**
+ * Utility function to transform a Cartesian coordinate into a hexagonal one.
+ *
+ * @param cartesian_x X-coordinate on the Cartesian grid.
+ * @param cartesian_y Y-coordinate on the Cartesian grid.
+ * @param cartesian_layout_height Height of the Cartesian layout.
+ * @param cartesian_z Z-coordinate on the Cartesian grid.
+ *
+ * @return coordinate on the hexagonal grid.
+ */
+template <typename Lyt>
+coordinate<Lyt> to_hex(int cartesian_x, int cartesian_y, int cartesian_layout_height, int cartesian_z)
+{
+    const auto y = cartesian_x + cartesian_y;
+    const auto x =
+        static_cast<int>(cartesian_x + std::ceil(std::floor(static_cast<double>(cartesian_layout_height) / 2) -
+                                                 static_cast<double>(y) / 2));
+    const auto z = cartesian_z;
+
+    fiction::coordinate<Lyt> hex{x, y, z};
+
+    return hex;
+}
+
+/**
+ * Transforms a 2DDWave-clocked Cartesian layout into a hexagonal even row clocked layout suitable for SiDBs by
+ * remapping all gates and wires.
+ *
+ * @param Lyt Gate-level layout that is 2DDWave-clocked.
+ *
+ * @return Hexagonal representation of the Cartesian layout.
+ */
+template <typename Lyt>
+hex_even_row_gate_clk_lyt hexagonalization(const Lyt& lyt)
+{
+    static_assert(is_gate_level_layout_v<Lyt>, "Lyt is not a gate level layout");
+    assert(lyt.is_clocking_scheme(clock_name::TWODDWAVE));
+
+    using hex_lyt = hex_even_row_gate_clk_lyt;
+
+    // get width & height of cartesian layout
+    const int layout_width  = static_cast<int>(lyt.x()) + 1;
+    const int layout_height = static_cast<int>(lyt.y()) + 1;
+
+    // calculate max width & height of hexagonal layout
+    const auto hex_height = to_hex<Lyt>(layout_width - 1, layout_height - 1, layout_height, 0).y;
+    const auto hex_width  = to_hex<Lyt>(layout_width - 1, 0, layout_height, 0).x;
+
+    // instantiate hexagonal layout
+    hex_lyt hex_layout{{hex_width, hex_height, 1}, fiction::row_clocking<hex_lyt>()};
+
+    // iterate through cartesian layout diagonally
+    for (int k = 0; k < layout_width + layout_height - 1; k++)
+    {
+        for (int x = 0; x < k + 1; x++)
+        {
+            const auto y = k - x;
+            if (y < layout_height && x < layout_width)
+            {
+                for (int z = 0; z < 2; z++)
+                {
+                    // old coordinate
+                    const coordinate<Lyt> old_coord{x, y, z};
+                    // new coordinate
+                    const coordinate<Lyt> hex{to_hex<Lyt>(x, y, layout_height, z)};
+
```

Comment:
You can significantly improve the runtime here. You are currently iterating over all coordinates whether there is something there or not. What you really want though is only the occupied coordinates. And ideally in topological ordering. To achieve this, you can simply wrap the layout in a `mockturtle::topo_view` and then call `foreach_node` on it.